langchain4j-redis/src/test/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStoreImplTest.java

```diff
@@ -0,0 +1,84 @@
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * disabled default, because this need local deployment of Redis
+ */
+@Disabled
+class RedisEmbeddingStoreImplTest {
+
+    private final EmbeddingStore<TextSegment> store = new RedisEmbeddingStoreImpl("http://localhost:6379", 4);
+
+    @Test
+    void testAdd() {
+        // test add without id
+        String id = store.add(Embedding.from(Arrays.asList(0.50f, 0.85f, 0.760f, 0.24f)),
+                TextSegment.from("test string", Metadata.metadata("field", "value")));
+        System.out.println("id=" + id);
+
+        // test add with id
+        String selfId = Utils.randomUUID();
+        store.add(selfId, Embedding.from(Arrays.asList(0.80f, 0.45f, 0.89f, 0.24f)));
+        System.out.println("id=" + selfId);
+    }
+
+    @Test
+    void testAddAll() {
+        // test add All Method without embedded
+        List<String> ids = store.addAll(Arrays.asList(
+                Embedding.from(Arrays.asList(0.3f, 0.87f, 0.90f, 0.24f)),
+                Embedding.from(Arrays.asList(0.54f, 0.34f, 0.67f, 0.24f)),
+                Embedding.from(Arrays.asList(0.80f, 0.45f, 0.779f, 0.5556f))
+        ));
+        System.out.println("ids=" + ids);
+
+        // test add all method with embedded
+        ids = store.addAll(Arrays.asList(
+                Embedding.from(Arrays.asList(0.3f, 0.87f, 0.90f, 0.24f)),
+                Embedding.from(Arrays.asList(0.54f, 0.34f, 0.67f, 0.24f)),
+                Embedding.from(Arrays.asList(0.80f, 0.45f, 0.779f, 0.5556f))
+        ), Arrays.asList(
+                TextSegment.from("testString1", Metadata.metadata("field1", "value1")),
+                TextSegment.from("testString2", Metadata.metadata("field2", "value2")),
+                TextSegment.from("testingString3", Metadata.metadata("field3", "value3"))
+        ));
+        System.out.println("ids=" + ids);
+    }
+
+    @Test
+    void testAddEmpty() {
+        // see log
+        store.addAll(Collections.emptyList());
+    }
+
+    @Test
+    void testFindRelevant() {
+        List<EmbeddingMatch<TextSegment>> res = store.findRelevant(Embedding.from(Arrays.asList(0.80f, 0.45f, 0.89f, 0.24f)), 5);
+        res.forEach(System.out::println);
+    }
+
+    @Test
+    void testScore() {
+        String id = store.add(Embedding.from(Arrays.asList(0.50f, 0.85f, 0.760f, 0.24f)),
+                TextSegment.from("test string", Metadata.metadata("field", "value")));
+        System.out.println("id=" + id);
+
+        // use the same embedding to search
+        List<EmbeddingMatch<TextSegment>> res = store.findRelevant(Embedding.from(Arrays.asList(0.50f, 0.85f, 0.760f, 0.24f)), 1);
+        res.forEach(System.out::println);
+
+        // the result embeddingMatch score is 5.96046447754E-8, but expected is 1 because they are same vectors.
```

Comment:
5.96046447754E-8 is 0,00000006 which is very close to 0. It is a cosine distance (in the range [0..2] where 0 is identical and 2 is opposite), not score(in the range [0..1] where 0 is opposite and 1 is identical). In the [documentation](https://redis.io/docs/interact/search-and-query/search/vectors/) they subtract distance from 1 to get the cosine similarity (in range [-1..1] where -1 is opposite and 1 is identical): `1 - distance`. We need to further convert similarity to score: `score = (2 - distance) / 2`

Reply:
> 5.96046447754E-8 is 0,00000006 which is very close to 0. It is a cosine distance (in the range [0..2] where 0 is identical and 2 is opposite), not score(in the range [0..1] where 0 is opposite and 1 is identical). In the [documentation](https://redis.io/docs/interact/search-and-query/search/vectors/) they subtract distance from 1 to get the cosine similarity (in range [-1..1] where -1 is opposite and 1 is identical): `1 - distance`. We need to further convert similarity to score: `score = (2 - distance) / 2`

I see. Thanks for your explanation! I changed the way to calculate right score by `score = (2 - distance) / 2`